controllers/datasource_controller.go

```diff
@@ -87,38 +102,71 @@ func (r *DatasourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-// Checkdatasource to update status
-func (r *DatasourceReconciler) Checkdatasource(ctx context.Context, logger logr.Logger, instance *arcadiav1alpha1.Datasource) error {
-	logger.Info("check datasource")
-	secret := corev1.Secret{}
-	if err := r.Get(ctx, types.NamespacedName{Namespace: instance.Namespace, Name: instance.Spec.AuthSecret}, &secret); err != nil {
-		return err
+func (r *DatasourceReconciler) Initialize(ctx context.Context, logger logr.Logger, instance *arcadiav1alpha1.Datasource) error {
+	instanceDeepCopy := instance.DeepCopy()
+	l := len(instanceDeepCopy.Finalizers)
+
+	var update bool
+
+	instanceDeepCopy.Finalizers = utils.AddString(instanceDeepCopy.Finalizers, arcadiav1alpha1.Finalizer)
+	if l != len(instanceDeepCopy.Finalizers) {
+		logger.V(1).Info("Add Finalizer for datasource", "Finalizer", arcadiav1alpha1.Finalizer)
+		update = true
 	}
-	accessKeyID := string(secret.Data[rootUser])
-	secretAccessKey := string(secret.Data[rootPassword])
-	endpoint := instance.Spec.URL
 
-	useSSL := false
+	if instanceDeepCopy.Labels == nil {
+		instanceDeepCopy.Labels = make(map[string]string)
+	}
 
-	// Initialize minio client object.
-	minioClient, err := minio.New(endpoint, &minio.Options{
-		Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
-		Secure: useSSL,
-	})
-	if err != nil {
-		return err
+	currentType := string(instanceDeepCopy.Spec.Type())
+	if v := instanceDeepCopy.Labels[arcadiav1alpha1.LabelDatasourceType]; v != currentType {
+		instanceDeepCopy.Labels[arcadiav1alpha1.LabelDatasourceType] = currentType
+		update = true
 	}
-	buckets, err := minioClient.ListBuckets(ctx)
-	if err != nil {
+
+	if update {
+		err := r.Client.Update(ctx, instanceDeepCopy)
+		if err != nil {
+			logger.Error(err, "")
+		}
```

Comment:
**Suggestion:** It seems like you are using a logger to log errors before returning them. This can lead to duplicate log entries if the caller also logs the error. It is a good practice to either log the error or return it, but not both. You should remove the logger calls before the error returns.
```suggestion
  err := r.Client.Update(ctx, instanceDeepCopy)
  return err
```